src/transforms/rule.ts

```diff
@@ -0,0 +1,219 @@
+// Copyright 2020 OpenStax Poland
+// Licensed under the MIT license. See LICENSE file in the project root for
+// full license text.
+
+import { Editor, Location, Node, Path, Range, Text, Transforms } from 'slate'
+
+import { Proof, Rule, RuleExample, RuleKind, Section, Statement } from '../interfaces'
+
+/**
+ * Create a new exercise and wrap currently selected block in it
+ *
+ * This function is similar to Slate's Transforms.wrapNodes, but it won't split
+ * line elements.
+ */
+export function insertRule(
+    editor: Editor,
+    kind?: RuleKind,
+    options: {
+        at?: Location,
+    } = {},
+): void {
+    Editor.withoutNormalizing(editor, () => {
+        const { at = editor.selection } = options
+
+        if (at == null) return
+
+        let [startPoint, endPoint] = Range.edges(Editor.range(editor, at))
+
+        // Find lowest non-inline nodes containing start and end points.
+        const [, startPath] = Editor.above(editor, {
+            at: startPoint,
+            match: n => !Text.isText(n) && !Editor.isInline(editor, n),
+        })
+        const [, endPath] = Editor.above(editor, {
+            at: endPoint,
+            match: n => !Text.isText(n) && !Editor.isInline(editor, n),
+        })
+
+        // If both points are in the same node we can defer to Slate's wrapNodes.
+        // Otherwise we need to do this manually, as wrapNodes with split: true
+        // doesn't work when match points at the editor.
```

Comment:
Can you extract this to an util function?

Reply:
Which part? (I'm asking because this comment highlights only three comment lines and no actual code)

Reply:
lines 51 - 86 - I think it may be used in more case than insering `rule`

Reply:
I tried extracting that into a new transform, and come to following conclusions:

1. that code depends on few assumptions checked, and variables declared, in lines 27 to 49. Extracting just lines 51 to 85 would result in a transform with API considerably different from other transforms;

2. `insertRule` ‘wraps’ nodes in a doubly nested element (76 - 85). This is something no other transform does. If we wanted to extract lines 51 - 86 that new transform probably should only use flat wrapper (like Slate's `Transforms.wrapNodes`), that however would require additional work in `insertRule`.

In short, while it is possible, I don't think it worth doing right now. We could certainly try it again if we ever need similar code somewhere else.